distsql: sorter memory accounting

RaduBerinde opened the pull request:
Switching `sorterValues` to use `RowContainer`.

Implementing all the primitives used by sorterstrategy was turning out to be
rather tedious; it was easier to simplify the sorterstrategy code: we now never
use a min-heap (we were doing heapsort which is slower than `sort.Sort`), and we
never `Push` or `Pop` on an existing heap.

As part of this change we are also plumbing `EvalContext` to `EncDatumRow`
comparisons (other than in the currently unused merge-joiner). Unfortunately
this is intertwined with the other change so I wasn't able to separate this in a
different commit.

RaduBerinde committed:
distsql: use RowContainer in sorter

Switching sorterValues to use RowContainer.

Implementing all the primitives used by sorterstrategy was turning out to be
rather tedious; it was easier to simplify the sorterstrategy code: we now never
use a min-heap (we were doing heapsort which is slower than `sort.Sort`), and we
never `Push` into an existing heap.

As part of this change we are also plumbing `EvalContext` to `EncDatumRow`
comparisons (other than in the currently unused merge-joiner). Unfortunately
this is intertwined with the other change so I wasn't able to separate this in a
different commit.

RaduBerinde committed:
distsql: remove Limit from SorterSpec

Thanks to `PostProcessSpec`, we can now figure out the limit directly without
help from the planner.

RaduBerinde committed:
sql: set fixed SQL memory limit to logic tests

We want to be able to test memory budged exceeded errors in logic tests, but
that is problematic if the memory limit varies by machine (at the very least,
the query will run longer on a larger machine).

Set a fixed limit of 128MB.

RaduBerinde committed:
distsql: add sorter memory accounting testcase

Add a test case that triggers a memory budget exceeded error due to a sorter.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15323)
<!-- Reviewable:end -->

RaduBerinde commented:
Made a small update, `EncRow` doesn't need to allocate each row (forgot that the output helper does that).

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 22 files at r1.
Review status: 0 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/sorter.go, line 90 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaSjiZTMZGTkHzmvk3:-KiaSjiZTMZGTkHzmvk4:b-ffptqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sorter.go#L90)):*
> ```Go
> 			ss = newSortAllStrategy(sv)
> 		} else {
> 			// No specified ordering match length but specified limit, we can optimize
> ```

nit: Since you're touching this line, please correct the wrong comma between the first 2 sentences. These break my eyes every time, although I also cherish them as I know exactly who wrote it.

---

*[pkg/sql/distsqlrun/sorterstrategy.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-Kia_3F8AzPRiIj2xRzR:-Kia_3F8AzPRiIj2xRzS:b-3ov8u9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sorterstrategy.go#L138)):*
> ```Go
> 		if int64(ss.sValues.Len()) < ss.k {
> 			// Accumulate up to k values.
> 			err = ss.sValues.AddRow(ctx, row)
> ```

consider breaking on err

---

*[pkg/sql/distsqlrun/sortervalues.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaWIj9p7T-oy4IiZrN:-KiaWIj9p7T-oy4IiZrO:bvqpews) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L33)):*
> ```Go
> // a sorter strategy, it is at this level that the rows to be sorted are stored.
> type sorterValues struct {
> 	sqlbase.RowContainer
> ```

nit: I probably wouldn't embed this. It's not like `sorterValues` can be transparently used as a container (or can it?)

---

*[pkg/sql/distsqlrun/sortervalues.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaWn7jSrxigdYKN_J_:-KiaWn7jSrxigdYKN_Ja:bxuxvlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L37)):*
> ```Go
> 	invertSorting bool // Inverts the sorting predicate.
> 	ordering      sqlbase.ColumnOrdering
> 	preallocRow   parser.Datums // Used temporarily.
> ```

nit: I don't know what "used temporarily" means. Like, is the field supposed to go away, or are its contents somehow ephemeral?
Also, I'd name it `preallocatedRow`, or `scratchRow/scratchSpace`.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaYRBtFJk7Jv1z7GLw:-KiaYRBtFJk7Jv1z7GLx:bbg4wv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L92)):*
> ```Go
> 	}
> 	// Avoid passing slice through interface{} to avoid allocation. We add the row
> 	// beforehand and sorterValues.Push does nothing.
> ```

nit: s/does nothing/is not used.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiabZEhjkA1eG-wry2H:-KiabZEhjkA1eG-wry2I:befuan9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L110)):*
> ```Go
> // ReplaceMax replaces the maximum element with the given row, if it is smaller.
> // Assumes InitMaxHeap was called.
> func (sv *sorterValues) ReplaceMax(row sqlbase.EncDatumRow) error {
> ```

consider `ReplaceMaxMaybe` or `MaybeReplaceMax`

---

*[pkg/sql/distsqlrun/stream_merger.go, line 132 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-Kiac6YWJF-r4IhH93iX:-Kiac6YWJF-r4IhH93iY:b-xtsrjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/stream_merger.go#L132)):*
> ```Go
> 
> 	// TODO(radu): plumb EvalContext
> 	evalCtx := &parser.EvalContext{}
> ```

I think we should try to do this plumbing, since maybe we'll get a real EvalCtx soon. No?

---

*[pkg/sql/testdata/logic_test/distsql_numtables, line 244 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiadshLw0OPbzoSLAn9:-KiadshLw0OPbzoSLAnA:b-41pez8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dccf42ce9fa080b32f86f4593473fded06ea2c46/pkg/sql/testdata/logic_test/distsql_numtables#L244)):*
> ```
> SELECT y, str, REPEAT('test', y) FROM NumToStr ORDER BY str
> 
> # Verify we use the "top K" strategy.
> ```

... and thus not OOM like above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-:-KiaeOGS93ae1Ckh9-Xm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 0 of 27 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/sorter.go, line 90 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaSjiZTMZGTkHzmvk3:-KiajFQF7QaJZ4OMulM9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sorter.go#L90)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: Since you're touching this line, please correct the wrong comma between the first 2 sentences. These break my eyes every time, although I also cherish them as I know exactly who wrote it.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/sorterstrategy.go, line 138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-Kia_3F8AzPRiIj2xRzR:-KiajM0KLGAtxMMVKbH8:bpahl2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sorterstrategy.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider breaking on err
</blockquote></details>

we exit below, but I moved the check in both branches, clearer this way.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaWIj9p7T-oy4IiZrN:-KiajT9ALXfRZ45ciUBC:b-11divn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L33)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I probably wouldn't embed this. It's not like `sorterValues` can be transparently used as a container (or can it?)
</blockquote></details>

It is, it's an augmented row container. I am actually planning to rename it to something to that effect so that other code (join) can use it.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 37 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaWn7jSrxigdYKN_J_:-KiajfMBEbAPEMWmBl9C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L37)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I don't know what "used temporarily" means. Like, is the field supposed to go away, or are its contents somehow ephemeral?
Also, I'd name it `preallocatedRow`, or `scratchRow/scratchSpace`.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 92 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiaYRBtFJk7Jv1z7GLw:-KiajjNpE3v3g7ijSLP4:b-vdrwrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L92)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/does nothing/is not used.
</blockquote></details>

this entire comment is bogus (left over), fixed.

---

*[pkg/sql/distsqlrun/sortervalues.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiabZEhjkA1eG-wry2H:-KiajvmxnfIAiJuQrYS7:b-qtmd3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/sortervalues.go#L110)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider `ReplaceMaxMaybe` or `MaybeReplaceMax`
</blockquote></details>

:musical_note: ..this is crazy.. :musical_note: 

---

*[pkg/sql/distsqlrun/stream_merger.go, line 132 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-Kiac6YWJF-r4IhH93iX:-Kiak3Y4h1PXT71LOjMk:bh1hin9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3326ed8067eacbbf78990c8f5e19703951481734/pkg/sql/distsqlrun/stream_merger.go#L132)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think we should try to do this plumbing, since maybe we'll get a real EvalCtx soon. No?
</blockquote></details>

We're not using mergejoins at all yet.

---

*[pkg/sql/testdata/logic_test/distsql_numtables, line 244 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-KiadshLw0OPbzoSLAn9:-KiakeIgGiIrvKgl6jrD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dccf42ce9fa080b32f86f4593473fded06ea2c46/pkg/sql/testdata/logic_test/distsql_numtables#L244)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

... and thus not OOM like above.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15323#-:-Kiakf84x-y4EiIHDm-Z:b-814g8k)*
<!-- Sent from Reviewable.io -->
